### PR TITLE
Gobierto Costs / Refactor code to the new dataset

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes https://github.com/PopulateTools/issues/issues/1116

## :v: What does this PR do?

- Adapt code to the new dataset, is now prepared to work with a year or multiple years.
- Some values in the dataset returns `undefined`, fix to convert these values.
- If the dataset only contains a single year hide `select` and `range-slider`.

## :mag: How should this be manually tested?
Staging

### CODE REVIEW

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 1]
```diff
@@ -105,7 +105,11 @@ export class CostsController {
 
     //Convert strings with some format to Numbers without format
     function convertStringToNumbers(amount) {
-      return Number(parseFloat(amount.replace(/\./g,'').replace(',','.')))
+      if (amount === '' || undefined) {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: `undefined` is being never evaluated. Upside down it does.
```suggestion
      if (amount === undefined || '') {
```

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 2]
```diff
@@ -115,31 +119,43 @@ export class CostsController {
...
+
       for (let amounts = 0; amounts < amountStrings.length; amounts++) {
         d[amountStrings[amounts]] = convertStringToNumbers(d[amountStrings[amounts]])
       }
+    }
 
-      d.ingressos = nanToZero(d.ingressos)
+    //Create an array with all years.
+    function getYearsFromData(arr, key) {
+      return [...new Map(arr.map(item => [item[key], item])).values()]
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: Are you trying to get _unique_ values? If so, replace `new Map` with `new Set`

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 3]
```diff
@@ -115,31 +119,36 @@ export class CostsController {
...
-    const totalData = [...data2018, ...data2019]
+    let yearsCosts = [...new Set(rawData.map(item => item.any))]
 
     let groupDataByYears = []
-    let groupData2018 = []
-    let groupData2019 = []
+
+    //Create an array of objects with all years
+    for (let index = 0; index < yearsCosts.length; index++) {
+      const data = rawData.filter(items => items.any == yearsCosts[index]).map(items => ({ ...items, population: population[index] }))
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: `filter` + `map` could be enhanced with `reduce`. Try this:
```suggestion
      const data = rawData.reduce((acc, item) => {
        if (item.any === yearsCosts[index]) {
          acc.push({ ...item, population: population[index] })
        }
        return acc
      }, [])
```
In such way, you run through the array once, whereas filter+map run through it twice
Don't forget the triple comparison

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 4]
```diff
@@ -115,31 +119,36 @@ export class CostsController {
...
+      const dataGroup = groupDataByYear(data, yearsCosts[index])
+      groupDataByYears.push(dataGroup)
+    }
 
     function groupDataByYear(data, year) {
-      //Filter groupData by "sumadatos" :https://github.com/PopulateTools/issues/issues/1097
-      let dataFilterSum = data.filter(element => element.sumadatos === '')
+      //Filter groupData by "sumadatos": https://github.com/PopulateTools/issues/issues/1097
+      let dataFilterSum = data.filter(element => element.suma_datos === '')
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: Destructure this

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 5]
```diff
@@ -164,19 +173,15 @@ export class CostsController {
 
         return r.set(key, item);
       }, new Map).values()];
-      return groupData = groupData.filter(element => element.agrupacio !== '' && element.year === year)
+      return groupData = groupData.filter(element => element.agrupacio !== '' && element.any === year)
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: Destructure this

File: app/javascript/gobierto_dashboards/modules/costs_controller.js [hunk 6]
```diff
@@ -164,19 +173,15 @@ export class CostsController {
...
     }
 
-    groupData2018 = groupDataByYear(data2018, '2018')
-    groupData2019 = groupDataByYear(data2019, '2019')
-
-    groupDataByYears = [...groupData2018, ...groupData2019]
-
-    groupDataByYears = groupDataByYears.sort((a, b) => (a.cost_total > b.cost_total) ? -1 : 1)
+    groupDataByYears = groupDataByYears.flat().sort((a, b) => (a.cost_total > b.cost_total) ? -1 : 1);
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/costs_controller.js]: Destructure always you can

File: config/locales/defaults/ca.yml [hunk 7]
```diff
@@ -15,7 +15,7 @@ ca:
     - dv
     - ds
     abbr_month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/ca.yml:18]: Typo?

File: config/locales/defaults/ca.yml [hunk 8]
```diff
@@ -43,7 +43,7 @@ ca:
       short: "%d %b"
       shortest: "%d %b %y"
     month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/ca.yml:46]: Typo?

File: config/locales/defaults/en.yml [hunk 9]
```diff
@@ -15,7 +15,7 @@ en:
     - Fri
     - Sat
     abbr_month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/en.yml:18]: Typo?

File: config/locales/defaults/en.yml [hunk 10]
```diff
@@ -43,7 +43,7 @@ en:
       short: "%b %d %Y"
       shortest: "%d %b %y"
     month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/en.yml:46]: Typo?

File: config/locales/defaults/es.yml [hunk 11]
```diff
@@ -15,7 +15,7 @@ es:
     - vie
     - sáb
     abbr_month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/es.yml:18]: Typo?

File: config/locales/defaults/es.yml [hunk 12]
```diff
@@ -43,7 +43,7 @@ es:
       short: "%d %b %Y"
       shortest: "%d %b %y"
     month_names:
-    -
+    -
```
Comment (Crashillo) [inline config/locales/defaults/es.yml:46]: Typo?

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: config/locales/defaults/ca.yml (hunk 7 above)
Comment (jorgeatgu) [reply config/locales/defaults/ca.yml:18]: > ferblape on 23 Jul Member
It looks like you added a space, but you can ignore it, we can't remove the blank here, it's a Rails default.

https://github.com/PopulateTools/gobierto/pull/3210/files?file-filters%5B%5D=.yml#r459266783

File: app/javascript/gobierto_dashboards/webapp/containers/costs/table/TableItem.vue [hunk 13]
```diff
@@ -429,22 +429,22 @@ export default {
   methods: {
     agrupacioData(id) {
       const yearFiltered = this.year
-      this.dataGroup = this.items.filter(element => element.codiact === id && element.year === yearFiltered)
+      this.dataGroup = this.items.filter(element => element.codiact === id && element.any === yearFiltered)
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/costs/table/TableItem.vue:432]: Oh lord!

**Review by Crashillo — APPROVED** [review]

**Pull request merged**